lib/vector-core/src/event/log_event.rs

```diff
@@ -520,7 +520,8 @@ impl LogEvent {
     /// or from the timestamp key set on the "Global Log Schema" (Legacy namespace).
     pub fn remove_timestamp(&mut self) -> Option<Value> {
         self.timestamp_path()
-            .and_then(|key| self.remove(key.as_str()))
+            .cloned()
+            .and_then(|key| self.remove(&key))
```

Comment:
Why the clone? It looks like we only use it as a reference. Same goes for most of the rest of the clones in the rest of the diff.

Reply:
Because `timestamp_path` needs immutable reference to `self` and then inside the closure `remove` requires a mutable reference to the same `self`. Same for the rest of the diff. 

Note that currently, we map/convert all paths to `String`s, see [here](https://github.com/vectordotdev/vector/blob/master/lib/vector-core/src/event/log_event.rs#L460) and [here](https://github.com/vectordotdev/vector/blob/master/lib/vector-core/src/event/log_event.rs#L317). in `LogEvent`. So this PR is at the very least avoiding the conversions. I didn't think about further optimizations at this point but always happy to hear more ideas.

Reply:
Ah, that makes sense. In this case it is actually safe because `timestamp` path really only needs an immutable reference to `self.metadata` while `remove` needs mutable access to `self.inner`, but the compiler can't see that through the function interfaces. We certainly don't need to do it as part of this change, but at some point it may be worth differentiating those so we can drop the clone.